### PR TITLE
Platform level: Reduce interface dispatch and trys in main loop

### PR DESCRIPTION
**Pull request by benaadams**
/cc @davidfowl @KrzysztofCwalina @sebastienros

### CODE REVIEW

**Comment by sebastienros** [comment]
6,460,818 RPS

**Comment by benaadams** [comment]
:)

**Comment by benaadams** [comment]
Added PR to change parsing to pass ReadOnlySequence via `in` https://github.com/aspnet/KestrelHttpServer/pull/2389

File: src/PlatformBenchmarks/HttpApplication.cs [hunk 1]
```diff
@@ -64,85 +67,90 @@ public virtual ValueTask OnReadCompletedAsync()
...
+                var consumed = buffer.Start;
+                var examined = buffer.End;
 
-                    var result = await task;
-                    var buffer = result.Buffer;
-                    var consumed = buffer.Start;
-                    var examined = buffer.End;
+                if (!buffer.IsEmpty)
+                {
+                    ParseHttpRequest(in buffer, out consumed, out examined);
```
Comment (davidfowl) [inline src/PlatformBenchmarks/HttpApplication.cs]: Why are you passing in here explicitly?

File: src/PlatformBenchmarks/HttpApplication.cs [hunk 2]
```diff
@@ -64,85 +67,90 @@ public virtual ValueTask OnReadCompletedAsync()
...
         }
 
-        private void ParseHttpRequest(ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
+        private void ParseHttpRequest(in ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
         {
             consumed = buffer.Start;
             examined = buffer.End;
 
-            if (_state == State.StartLine)
+            var parsedStartLine = _state == State.StartLine;
```
Comment (davidfowl) [inline src/PlatformBenchmarks/HttpApplication.cs]: parsingStartLine?

File: src/PlatformBenchmarks/HttpApplication.cs (hunk 1 above)
Comment (davidfowl) [inline src/PlatformBenchmarks/HttpApplication.cs]: Why are you specifying in.

Comment (benaadams) [reply src/PlatformBenchmarks/HttpApplication.cs]: No reason; will change :)

File: src/PlatformBenchmarks/HttpApplication.cs (hunk 2 above)
Comment (benaadams) [reply src/PlatformBenchmarks/HttpApplication.cs]: Changed

File: src/PlatformBenchmarks/HttpApplication.cs [hunk 3]
```diff
@@ -64,85 +67,90 @@ public virtual ValueTask OnReadCompletedAsync()
...
+        private void ParseHttpRequest(in ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
         {
             consumed = buffer.Start;
             examined = buffer.End;
 
-            if (_state == State.StartLine)
+            var parsingStartLine = _state == State.StartLine;
+            if (parsingStartLine)
             {
                 if (Parser.ParseRequestLine(this, buffer, out consumed, out examined))
```
Comment (KrzysztofCwalina) [inline src/PlatformBenchmarks/HttpApplication.cs]: Would it make sense to allow passing buffer by-in?

Comment (davidfowl) [reply src/PlatformBenchmarks/HttpApplication.cs]: Yes, we just changed this in Kestrel.

**Comment by ErikEJ** [comment]
Now go fix SqlClient 😁

**Comment by sebastienros** [comment]
@ErikEJ yeah, and we are actively working on making it faster too, but FYI SqlClient is already faster than the implementations for NodeJs and Java that we tested, measured on Windows (SNI) and Linux (managed). The SqlClient team has already made great improvements that will be visible with the next release (faster on Linux, faster azure encrypted connections, reliability, ...), and it will also benefit 2.0 apps.

**Pull request merged**